watchtower/wtwire/wtwire_test.go

```diff
@@ -0,0 +1,185 @@
+package wtwire_test
+
+import (
+	"bytes"
+	"math/rand"
+	"reflect"
+	"testing"
+	"testing/quick"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/watchtower/wtwire"
+)
+
+// TestWatchtowerWireProtocol uses the testing/quick package to create a series
+// of fuzz tests to attempt to break a primary scenario which is implemented as
+// property based testing scenario.
+func TestWatchtowerWireProtocol(t *testing.T) {
+	t.Parallel()
+
+	// mainScenario is the primary test that will programmatically be
+	// executed for all registered wire messages. The quick-checker within
+	// testing/quick will attempt to find an input to this function, s.t
+	// the function returns false, if so then we've found an input that
+	// violates our model of the system.
+	mainScenario := func(msg wtwire.Message) bool {
+		// Give a new message, we'll serialize the message into a new
+		// bytes buffer.
+		var b bytes.Buffer
+		if _, err := wtwire.WriteMessage(&b, msg, 0); err != nil {
+			t.Fatalf("unable to write msg: %v", err)
+			return false
+		}
+
+		// Next, we'll ensure that the serialized payload (subtracting
+		// the 2 bytes for the message type) is _below_ the specified
+		// max payload size for this message.
+		payloadLen := uint32(b.Len()) - 2
+		if payloadLen > msg.MaxPayloadLength(0) {
+			t.Fatalf("msg payload constraint violated: %v > %v",
+				payloadLen, msg.MaxPayloadLength(0))
+			return false
+		}
+
+		// Finally, we'll deserialize the message from the written
+		// buffer, and finally assert that the messages are equal.
+		newMsg, err := wtwire.ReadMessage(&b, 0)
+		if err != nil {
+			t.Fatalf("unable to read msg: %v", err)
+			return false
+		}
+		if !reflect.DeepEqual(msg, newMsg) {
+			t.Fatalf("messages don't match after re-encoding: %v "+
+				"vs %v", spew.Sdump(msg), spew.Sdump(newMsg))
+			return false
+		}
+
+		return true
+	}
+
+	type typeGenFunc func([]reflect.Value, *rand.Rand)
+
+	// customTypeGen is a map of functions that are able to randomly
+	// generate a given type. These functions are needed for types which
+	// are too complex for the testing/quick package to automatically
+	// generate.
+	customTypeGen := map[wtwire.MessageType]typeGenFunc{
+		wtwire.MsgSessionInit: func(v []reflect.Value, r *rand.Rand) {
```

Comment:
Due to the simplicity of many of these messages, I think we can actually leave most of these off all together. These are only required if the struct has complex fields like a `*btcec.PublicKey`. I'd try to remove them all, the only add them back due to panics in the internal reflect code for `testing/quick`.

Reply:
TIL, thanks!